src/transform/src/predicate_pushdown.rs

```diff
@@ -458,6 +454,120 @@ impl PredicatePushdown {
                     .or_insert_with(HashSet::new)
                     .clear();
             }
+            MirRelationExpr::Join {
+                inputs,
+                equivalences,
+                ..
+            } => {
+                expr::canonicalize::sort_dedup_equivalences(equivalences);
+                // Push down:
+                // 1) equality constraints supported by the same single
+                //    input.
+                // 2) equality constraints consisting of an expression
+                //    supported by a single input + a literal
+
+                let input_mapper = expr::JoinInputMapper::new(inputs);
+                // Predicates to push at each input.
+                let mut push_downs = vec![Vec::new(); inputs.len()];
+
+                for equivalence in equivalences.iter_mut() {
+                    if equivalence.iter().filter(|expr| expr.is_literal()).count() > 1 {
+                        // Because of equivalences have been dedupped, this
+                        // means that everything in the equivalence class must
+                        // be equal to two different literals, so the entire
+                        // relation zeroes out
+                        relation.take_safely();
+                        return;
+                    }
+
+                    // Go through the equivalence class and remove an expression
+                    // if the expression is single-input, and there is another
+                    // expression still in the equivalence class that belongs to
+                    // no inputs or the same input.
```

Comment:
I don't understand most of the logic in the block below. I could imagine it's all on-point, but the more complicated expressions being built up might benefit from an explanation in text-y form.

Reply:
Will expand explanation/attempt simplification. Also I noticed the block contains a bit of detritus from trying to debug the dbt test failure.